File: frontend/scripts/blog_author.py
from copy import deepcopy

AUTHOR_ID = "finboard-team"
AUTHOR_NAME = "FinBoard Team"
SCHEMA_AUTHOR = {
    "@type": "Organization",
    "@id": "https://finboard.ai/#organization",
    "name": AUTHOR_NAME,
    "url": "https://finboard.ai/about",
}
ARTICLE_TYPES = {
    "Article",
    "AdvertiserContentArticle",
    "NewsArticle",
    "AnalysisNewsArticle",
    "AskPublicNewsArticle",
    "BackgroundNewsArticle",
    "OpinionNewsArticle",
    "ReportageNewsArticle",
    "ReviewNewsArticle",
    "Report",
    "SatiricalArticle",
    "ScholarlyArticle",
    "MedicalScholarlyArticle",
    "SocialMediaPosting",
    "BlogPosting",
    "LiveBlogPosting",
    "DiscussionForumPosting",
    "TechArticle",
    "APIReference",
}


def _schema_type_name(value):
    if not isinstance(value, str):
        return None
    for prefix in ("http://schema.org/", "https://schema.org/"):
        if value.startswith(prefix):
            return value[len(prefix):]
    return value


def _is_article_type(value):
    values = value if isinstance(value, list) else [value]
    return any(_schema_type_name(item) in ARTICLE_TYPES for item in values)


def normalize_structured_data(value):
    if isinstance(value, list):
        return [normalize_structured_data(item) for item in value]
    if not isinstance(value, dict):
        return value

    normalized = {
        key: normalize_structured_data(child)
        for key, child in value.items()
    }
    if _is_article_type(normalized.get("@type")):
        normalized["author"] = deepcopy(SCHEMA_AUTHOR)
    return normalized


def normalize_post(post):
    normalized = deepcopy(post)
    normalized["author"] = AUTHOR_NAME
    normalized["authorId"] = AUTHOR_ID
    normalized["structuredData"] = normalize_structured_data(
        normalized.get("structuredData")
    )
    return normalized

File: frontend/src/lib/blogAuthor.mjs
export const FINBOARD_TEAM_AUTHOR = Object.freeze({
  id: "finboard-team",
  name: "FinBoard Team",
  role: "",
  bio: "Written by the FinBoard team. We build multi-entity reporting and consolidation for QuickBooks Online, and we write from what we see across our customers' books.",
  linkedin: "https://finboard.ai/about",
});

export const FINBOARD_TEAM_SCHEMA_AUTHOR = Object.freeze({
  "@type": "Organization",
  "@id": "https://finboard.ai/#organization",
  name: "FinBoard Team",
  url: "https://finboard.ai/about",
});

const ARTICLE_TYPES = new Set([
  "Article",
  "AdvertiserContentArticle",
  "NewsArticle",
  "AnalysisNewsArticle",
  "AskPublicNewsArticle",
  "BackgroundNewsArticle",
  "OpinionNewsArticle",
  "ReportageNewsArticle",
  "ReviewNewsArticle",
  "Report",
  "SatiricalArticle",
  "ScholarlyArticle",
  "MedicalScholarlyArticle",
  "SocialMediaPosting",
  "BlogPosting",
  "LiveBlogPosting",
  "DiscussionForumPosting",
  "TechArticle",
  "APIReference",
]);

function schemaTypeName(value) {
  if (typeof value !== "string") return null;
  return value.replace(/^https?:\/\/schema\.org\//, "");
}

function isArticleType(value) {
  const types = Array.isArray(value) ? value : [value];
  return types.some((type) => ARTICLE_TYPES.has(schemaTypeName(type)));
}

export function resolveAuthor() {
  return FINBOARD_TEAM_AUTHOR;
}

export function normalizeBlogStructuredData(value) {
  if (Array.isArray(value)) {
    return value.map(normalizeBlogStructuredData);
  }

  if (!value || typeof value !== "object") {
    return value;
  }

  const normalized = Object.fromEntries(
    Object.entries(value).map(([key, child]) => [
      key,
      normalizeBlogStructuredData(child),
    ]),
  );

  if (isArticleType(normalized["@type"])) {
    normalized.author = { ...FINBOARD_TEAM_SCHEMA_AUTHOR };
  }

  return normalized;
}
